src/clj/witan/send/check_inputs.clj

```diff
@@ -28,15 +29,54 @@
       (set all-states)
       (set (map #(dissoc % :academic-year) all-states)))))
 
+(defn joiner? [age transition]
+  (and (= (:academic-year-2 transition) age) (= (:setting-1 transition) :NONSEND)))
+
+(defn leaver? [age transition]
+  (and (= (:academic-year-1 transition) age) (= (:setting-2 transition) :NONSEND)))
+
+(defn check-all-ages-present [transitions]
+  "Checks all ages are represented in both academic years 1 and 2"
+  (let [transitions-to-ages (distinct (map :academic-year-2 transitions))
+        transitions-from-ages (distinct (map :academic-year-1 transitions))
+        all-ages (sort (distinct (concat transitions-to-ages transitions-from-ages)))
+        possible-ages (range (first all-ages) (+ 1 (last all-ages)))]
+    (doseq [a (drop 1 possible-ages)]
+      (when-not (some #(= a %) transitions-to-ages)
+        (do (r/info (r/bold (str "There are no transitions to AY " a)))
+            (repl-warn (str "There are no transitions to AY " a)))))
+    (doseq [a (drop-last 1 possible-ages)]
+      (when-not (some #(= a %) transitions-from-ages)
+        (do (r/info (r/bold (str "There are no transitions from AY " a)))
+            (repl-warn (str "There are no transitions from AY " a)))))))
```

Comment:
With the 2 doseqs above, can we separate the mapping of the predicates over the data and the updating of the report/repl printing so that we can test that doing something like `(keep (some #(= a %) transitions-to-ages) (drop 1 possible-ages)` does what we expect it to do?

Reply:
Are you suggesting we just map twice, in two places, to meet this requirement?